File: src/main/java/database/FamilyTable.java
package database;

import com.koenig.commonModel.Family;
import com.koenig.commonModel.User;
import com.koenig.commonModel.database.DatabaseItem;
import com.koenig.commonModel.database.UserService;
import com.koenig.communication.messages.FamilyMessage;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.List;

public class FamilyTable extends Table<Family> {
    public static final String NAME = "family_table";
    public static final String USERS = "users";
    private final UserService userService;

    public FamilyTable(Connection connection, UserService userService) {
        super(connection);
        this.userService = userService;
    }

    @Override
    public String getTableName() {
        return NAME;
    }

    @Override
    protected Family getItem(ResultSet rs) throws SQLException {
        String usersText = rs.getString(USERS);
        List<User> users = getUsers(userService, usersText);
        String family = rs.getString(COLUMN_NAME);
        return new Family(family, users);
    }

    @Override
    protected String getTableSpecificCreateStatement() {
        return ", " + USERS + " TEXT ";
    }

    @Override
    protected void setItem(NamedParameterStatement ps, Family item) throws SQLException {
        setUsers(item.getUsers(), ps, USERS);
    }

    @Override
    protected List<String> getColumnNames() {
        return Arrays.asList(USERS);
    }

    public void addUserToFamily(Family family, String userId) throws SQLException {
        lock.lock();
        try {
            String selectQuery = "UPDATE " + getTableName() + " SET " + getNamedParameter(USERS) + " WHERE " + getNamedParameter(COLUMN_ID);

            NamedParameterStatement statement = new NamedParameterStatement(connection, selectQuery);
            statement.setString(COLUMN_ID, family.getId());
            StringBuilder builder = new StringBuilder();
            for (User user : family.getUsers()) {
                builder.append(user.getId());
                builder.append(FamilyMessage.SEPARATOR);
            }

            builder.append(userId);
            String users = builder.substring(0, builder.length());
            statement.setString(USERS, users);
            statement.executeUpdate();
        } finally {
            lock.unlock();
        }
    }

    public Family getFamilyByName(String familyName) throws SQLException {
        lock.lock();
        try {
            DatabaseItem<Family> family = null;
            String selectQuery = "SELECT * FROM " + getTableName() + " WHERE " + COLUMN_DELETED + " = :" + COLUMN_DELETED + " AND " + getNamedParameter(COLUMN_NAME);

            NamedParameterStatement statement = new NamedParameterStatement(connection, selectQuery);
            statement.setInt(COLUMN_DELETED, FALSE);
            statement.setString(COLUMN_NAME, familyName);
            ResultSet rs = statement.executeQuery();
            while (rs.next()) {
                family = resultToItem(rs);
            }

            return family == null ? null : family.getItem();
        } finally {
            lock.unlock();
        }
    }
}

File: src/main/java/database/FamilyTable.kt
package database

import com.koenig.commonModel.Family
import com.koenig.commonModel.Repository.FamilyRepository
import com.koenig.commonModel.User
import com.koenig.commonModel.database.DatabaseItem
import com.koenig.commonModel.database.DatabaseItemTable
import com.koenig.commonModel.toLocalDate
import com.koenig.commonModel.toYearMonth
import com.koenig.communication.messages.FamilyMessage
import org.joda.time.LocalDate
import org.joda.time.YearMonth
import java.sql.Connection
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*

class FamilyTable(connection: Connection, private val userService: (String) -> User?) : ItemTable<Family>(connection) {

    override val tableName: String
        get() = NAME

    override val itemSpecificCreateStatement: String
        get() = ", $USERS TEXT , $START_MONTH INT"

    override val columnNames: List<String>
        get() = Arrays.asList(USERS, START_MONTH)

    @Throws(SQLException::class)
    override fun getItem(rs: ResultSet): Family {
        val usersText = rs.getString(USERS)
        val users = getUsers(userService, usersText)
        val startDate = rs.getYearMonth(START_MONTH)
        val family = rs.getString(DatabaseItemTable.COLUMN_NAME)
        return Family(family, users, startDate)
    }

    @Throws(SQLException::class)
    override fun setItem(ps: NamedParameterStatement, item: Family) {
        setUsers(item.users, ps, USERS)
        ps.setYearMonth(START_MONTH, item.startMonth)
    }

    @Throws(SQLException::class)
    fun addUserToFamily(family: Family, userId: String) {
        lock.lock()
        try {
            val selectQuery = "UPDATE " + tableName + " SET " + getNamedParameter(USERS) + " WHERE " + getNamedParameter(DatabaseItemTable.COLUMN_ID)

            val statement = NamedParameterStatement(connection, selectQuery)
            statement.setString(DatabaseItemTable.COLUMN_ID, family.id)
            val builder = StringBuilder()
            for (user in family.users) {
                builder.append(user.id)
                builder.append(FamilyMessage.SEPARATOR)
            }

            builder.append(userId)
            val users = builder.substring(0, builder.length)
            statement.setString(USERS, users)
            statement.executeUpdate()
        } finally {
            lock.unlock()
        }
    }

    @Throws(SQLException::class)
    fun getFamilyByName(familyName: String): Family? {
        lock.lock()
        try {
            var family: DatabaseItem<Family>? = null
            val selectQuery = "SELECT * FROM " + tableName + " WHERE " + DatabaseItemTable.COLUMN_DELETED + " = :" + DatabaseItemTable.COLUMN_DELETED + " AND " + getNamedParameter(DatabaseItemTable.COLUMN_NAME)

            val statement = NamedParameterStatement(connection, selectQuery)
            statement.setInt(DatabaseItemTable.COLUMN_DELETED, DatabaseItemTable.FALSE)
            statement.setString(DatabaseItemTable.COLUMN_NAME, familyName)
            val rs = statement.executeQuery()
            while (rs.next()) {
                family = getDatabaseItem(rs)
            }

            return family?.item
        } finally {
            lock.unlock()
        }
    }

    companion object {
        val NAME = "family_table"
        val USERS = "users"
        val START_MONTH = "start_month"
    }
}

fun ResultSet.getYearMonth(key: String): YearMonth {
    return getInt(key).toYearMonth()
}

fun ResultSet.getLocalDate(key: String): LocalDate {
    return getInt(key).toLocalDate()
}

class FamilyDbRepository(val table: FamilyTable) : FamilyRepository {
    override val allFamilies: List<Family>
        get() = table.allItems

}
